keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,233 @@
+becomes the kubernetes-csi release version, which is increased
+according to the same rules as the individual components (major
+version bump when any of its components had a major change, etc.).
+
+#### Release artifacts
+
+Tagging a component with a semantic version number triggers a release
+build for that component. The output is primarily the container image
+for components that need to be deployed. Those images get published as
+`gcr.io/kubernetes-csi/<image>:<tag>` pending [issue
+158](https://github.com/kubernetes/k8s.io/issues/158).
+
+Only binaries provided as part of such a release should be considered
+production ready. Binaries are never going to be rebuilt, therefore an
+image like `csi-node-driver-registrar:v1.0.2` will always pull the
+same content and `imagePullPolicy: Always` can be omitted.
+
+Eventually auxiliary files like the RBAC rules that are included in
+each component might also get published in a single location, like
+`https://dl.k8s.io/kubernetes-csi/`, but this currently outside of
+this KEP.
+
+#### Release process
+
+* A change is submitted against master.
+
+* A [Prow](https://github.com/kubernetes/test-infra/blob/master/prow/README.md)
+  job checks out the source of a modified component, rebuilds it and then
+  runs unit tests and E2E tests with it.
+
+* Maintainers accept changes into the master branch.
+
+* The same Prow job runs for master and repeats the check. If it succeeds,
+  a new "canary" image is published for the component.
+
+* In preparation for the release of a major new update, a feature freeze is
+  declared for the "master" branch and only changes relevant for that next
+  release are accepted.
+
+* When all changes targeted for the release are in master, automatic
+  test results are okay and and potentially some more manual tests,
+  maintainers tag each component directly on the master branch.
+
+* Maintenance releases are prepared by branching a "release-X.Y" branch from
+  release "vX.Y" and backporting relevant fixes from master. The same
+  prow job as for master also handles the maintenance branches, but potentially
+  with a different configuration.
+
+
+### Implementation Details
+
+Each component has its own release configuration (what to build and
+publish) and rules (scripts, makefile). The advantage is that those
+can be branched and tagged together with the component.
+
+To simplify maintenance and ensure consistency, the common parts can
+be shared via
+[csi-release-tools](https://github.com/kubernetes-csi/csi-release-tools/).
+
+The prow job then just provides a common execution environment, with
+the ability to bring up test cluster(s), publish container images on
+and other files.
+
+### Risks and Mitigations
+
+Pushing a new release image is triggered by setting a tag. Unless
+there is a build failure, the result of the automatic build becomes
+the latest release, without any further manual checking. There are
+multiple risks:
```

Comment:
Can we eliminate this risk by adding one more step of doing complete hostpath e2e tests with the latest canary?  And/or tagging an "rc" build and use that image for e2e validation?

Reply:
> Can we eliminate this risk by adding one more step of doing complete hostpath e2e tests with the latest canary?

And then what? Who turns the "canary" into "release"? In the centralized release process, each component gets tagged (in its own repo) but then building and pushing production images is part of the responsibilities of the central release job.

In the decentralized model we don't have such a job.

> And/or tagging an "rc" build and use that image for e2e validation?

The testing of the branch already covers that. I need to be clearer about that, but the intention ("job checks out the source of a modified component") was that the latest code of the component gets tested, not the lastest tagged code.

 So if a component maintainer wants to know whether the latest revision works, they just need to look at those test results. The risk is not that we don't test before releasing - it's rather that maintainers can accidentally ignore those results.

Reply:
Let me add that the proposal actually has one round of "test released images" before the "announce a kubernetes-csi release": that's the testing that happens in csi-driver-host-path when building a tagged csi-driver-host-path release.

In the GitHub UI, we could tag a "v1.2.0" and mark it as "pre-release". That'll trigger a build for the v1.2.0 tag. Then when that build has completed, we will have all the additional release artifacts for that version (none at the moment, but perhaps we'll add something eventually, like a snapshot of the RBAC files) and we can edit the release in GitHub to remove the "pre-release" flag.

Reply:
I would prefer to tag a commit with an "rc" if we still intend to do more testing/validation.  Once all the validation is done, then you can create an official release at the same commit as the rc.

With the other way, you could tag "v1.2.0", find a bug in testing, check in more changes, but now you have to tag "v1.2.1", so you end up skipping version numbers.

Reply:
It's okay to tag as "rc". Conceptually it is identical to testing of the canary images, but referencing the tested commits becomes easier.

> With the other way, you could tag "v1.2.0", find a bug in testing, check in more changes, but now you have to tag "v1.2.1", so you end up skipping version numbers.

Strictly speaking, v1.2.0 wouldn't be skipped. It's still there, it just didn't pass testing and directly gets superseded by the next version. But I get what you mean.